src/trail.js

```diff
@@ -0,0 +1,237 @@
+        var i,
+            n = data.length, 
+            d,
+            def,
+            s,
+            defined0 = false,
+            defined1 = false,
+            buffer;
+
+        //make global optimization for radius when r1 + r2 > lxy
+        for(i = 0; i < n; i++){
+            def = defined(d = data[i], i, data) && (s = +size(d, i, data));
+            if(!(i < n && def) === defined1){
+                if (defined1 = !defined1) ready2 = 0;
+            }
+            if(defined1) {
+                scaleRatio(+x(d, i, data), +y(d, i, data), s);
+            }
+        }
+
+        if(context == null) context = buffer = path();
+
+        var start,
+            j,
+            d1,
+            def1,
+            s1,
+            tag = false;
+
+        for(i = 0; i < n; i++){
+            def = defined(d = data[i], i, data) && (s = +size(d, i, data));
+            if(!(i < n && def) === defined0){
+                if(defined0 = !defined0){
+                    ready = 0;
+                    start = i;
+                }
+            }
+            if(defined0){
+                j = i + 1;
+                if(j < n){
+                    def1 = defined(d1 = data[j], j, data) && (s1 = +size(d1, j, data));
+                    if(def1){
+                        segment(+x(d, i, data), +y(d, i, data), s*scaleRatio_min, +x(d1, j, data), +y(d1, j, data), s1*scaleRatio_min);
+                    }
+                    else{
+                        tag = true;
+                    }
+                }
+                else{
+                    tag = true;
+                }
+                if(tag){
+                    j = i - 1;
+                    if(j < start){
+                    var tx = +x(d, i, data),
+                        ty = +y(d, i, data),
+                        ts = +size(d, i, data);
+                    
+                    context.moveTo(tx + ts*scaleRatio_min / 2, ty);
+                    context.arc(tx, ty, ts*scaleRatio_min / 2, 0, tau, 0);
+                    context.closePath();
+                    }
+                    else{
+                        while(j >= start){
+                            d1 = data[j];
+                            segment(+x(d, j + 1, data), +y(d, j + 1, data), +size(d, j + 1, data)*scaleRatio_min, +x(d1, j, data), +y(d1, j, data), +size(d1, j, data)*scaleRatio_min);
+                            d = d1;
+                            j -= 1;
+                        }
+                        j += 1;
+                        d1 = data[j + 1];
+                        segment(+x(d, j, data), +y(d, j, data), +size(d, j, data)*scaleRatio_min, +x(d1, j + 1, data), +y(d1, j + 1, data), +size(d1, j + 1, data)*scaleRatio_min);
+
+                        context.closePath();
+                    }
+                    tag = false;
+                }
+            }
+        }
+
+        if(buffer){
+            context = null;
+            return buffer + '' || null;
```

Comment:
when is buffer truthy and buffer+'' falsy?

Reply:
when data = []，buffer(object) is truthy but buffer+''(string) is falsy.